### DISCUSSION TITLE
bug: cannot chain query or mutation across files

### DISCUSSION THREAD

**Discussion by ollyde** [trpc/trpc#2581] [Q&A] [answered]
### Provide environment information

System:
    OS: macOS 12.5.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 11.72 GB / 64.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
  Browsers:
    Chrome: 104.0.5112.101
    Firefox: 103.0.2
    Safari: 15.6.1

### Describe the bug

Cannot join routers across files, making organization useless.

In the example 'getUsers' only works if it's in the same file.

### To reproduce

**Index**

```
import * as trpc from '@trpc/server';
import { Context } from 'vm';
import { getUser } from './get-user';

// Works
// let appRouter = trpc.router();
// .query('getUser', {
//   async resolve() {
//     const request = (await User.findOne({ where: { id: 'test' } }))!;
//     return request;
//   },
// });

// "Endpoints", the name is the same as the call but with starting lower case.
// For example CreateUser -> createUser

const createRouter = () => {
  return trpc.router<Context>();
};

let userRouter = createRouter();

// User
getUser(userRouter);

export const initTrpc = () => {
  // Empty init to call the file.
};

const appRouter = createRouter();
appRouter.merge(userRouter);

export { appRouter };

export type AppRouter = typeof appRouter;
```

**getUsers.ts is broken**

```
import { Context } from 'vm';
import User from '../models/user';
import * as trpc from '@trpc/server';
import { Router } from '@trpc/server/dist/declarations/src/router';

export const initGetUser = (appRouter) => {
  // appRouter.query('getUser', {
  //   input: z.object({
  //     id: z.string(),
  //   }),
  //   async resolve({ input }) {
  //     const user = (await User.findOne({ where: { id: input.id } }))!;
  //     return user;
  //   },
  // });
};

export const getUser = (
  router: Router<Context, Context, {}, {}, {}, {}, trpc.DefaultErrorShape>,
) => {
  return router.query('getUser', {
    async resolve() {
      const request = (await User.findOne({ where: { id: 'test' } }))!;
      return request;
    },
  });
};
```

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by ollyde** [comment]
I'd also like to not that separating functions into files isn't working. Seems to be no way to separate each path into it's own file without creating 100s of routers.

**Comment by ollyde** [comment]
<img width="159" alt="Screenshot 2022-08-28 at 22 13 36" src="https://user-images.githubusercontent.com/6940726/187092771-253df71d-6d13-4848-9132-41c80734cbd4.png">

For example, how would one do this?

**Comment by KATT** [comment, marked as answer]
See here for working reference examples:

- https://trpc.io/docs/v9/merging-routers#example-code
- https://github.com/trpc/examples-next-prisma-starter/tree/main/src/server/routers
- https://github.com/calcom/cal.com/tree/main/packages/trpc/server/routers

**Reply by ollyde** [reply]
@KATT I’ve seen the examples and they don’t work. Is something missing? Would be nice if you converted this to discussion after I replied. This is still a bug; perhaps you’d like the project link?

**Reply by ollyde** [reply]
@KATT also there seems no way to separate the functions; which is fairly standard API layout 🤪 the examples show large router files which isn’t great; also running the examples locally; it doesn’t actually work..

**Reply by KATT** [reply]
Thousands of applications run on tRPC, it might use a pattern (the builder pattern) that you're not used to, but rest assured, it does work.

Create a repository where it's not working and I'm happy to do a PR to show you how to use it.

**Reply by KATT** [reply]
As an example, in your code you write

```ts
const appRouter = createRouter();
appRouter.merge(userRouter);
```

When it should be

```ts

const appRouter = createRouter()
  .merge(userRouter);
```

**Reply by juliusmarminge** [reply]
> @KATT also there seems no way to separate the functions; which is fairly standard API layout 🤪 the examples show large router files which isn’t great; also running the examples locally; it doesn’t actually work..

If you want to do separation, just separate the input and resolver: 
```ts
// anywhere
const getUserValidator = z.object({ ... });
const transformUsers = (users: User[]) => { ... };

// routers/user.ts
const userRouter = createRouter()
  .query("getUsers", {
    input: getUserResolver,
    async resolve({ input, ctx }) {
      const users = ctx.prisma.user.findMany(...);
      return transformUsers(users);
    }
  });

// routers/index.ts
const appRouter = createRouter()
  .merge(userRouter);
```

**Reply by KATT** [reply]
FWIW, you might want to check out the v10 alpha too - https://trpc.io/docs/v10/

**Reply by ollyde** [reply]
> As an example, in your code you write
> 
> ```ts
> const appRouter = createRouter();
> appRouter.merge(userRouter);
> ```
> 
> When it should be
> 
> ```ts
> const appRouter = createRouter()
>   .merge(userRouter);
> ```

Thanks for the replies.

So it’s indeed a bug; the chaining is broken and there’s no way to separate each call into a separate file which in our eyes makes the library useless. Routers could be 1000s of lines for example. Just feedback; think we’ll wait until these bugs are basic feature are implemented 😎

**Reply by ollyde** [reply]
@KATT 10 is looking good! Just hope we can separate them into different files :-)

**Reply by juliusmarminge** [reply]
> @KATT 10 is looking good! Just hope we can separate them into different files :-)

You can indeed

**Reply by ollyde** [reply]
@juliusmarminge I'm looking at the https://trpc.io/docs/v9/authorization just a tip, since it's near the same subject here, would be nice to apply this 'per procedure' and not to all of them in the same route.

**Reply by KATT** [reply]
Nothing is broken and it's totally possible to do 1 procedure per file if that's what you like. You can also nest routers infinitely. Use your creativity within the constraints of the tool, contribute, or make your own tool if this is so "useless".

Here's an example of doing 1 proc per file:

- https://github.com/trpc/trpc/tree/katt/showcase-2581/examples/next-prisma-starter/src/server/routers
- https://github.com/trpc/trpc/pull/2588

First and only warning: please be respectful & stop calling your feature request "a bug": 

- you have not contributed to tRPC in any shape or form
- you are not suggesting any feasible API design on how such a thing would look like nor are you being constructive with the feedback
